### PR TITLE
adding PINN to the pipeline. 

### LINKED ISSUE #65: Integrating pinn on pyLOM
The goal of this issue is to integrate Physics informed neural networks [https://arxiv.org/abs/1711.10561](https://arxiv.org/abs/1711.10561) on the pipeline

### PR DESCRIPTION
**Pull request by DavidRamosArchilla**
Closes #65

### CODE REVIEW

**Comment by ArnauMiro** [comment]
@DavidRamosArchilla what is the status of the PINN? Can we start a revision and add it to develop or shall we wait still (if so it might be wise to convert this PR to draft)?

**Comment by DavidRamosArchilla** [comment]
> @DavidRamosArchilla what is the status of the PINN? Can we start a revision and add it to develop or shall we wait still (if so it might be wise to convert this PR to draft)?

I would consider the code of the PINN finished, from this version i would just need to change some minor thing that we modified in our version. I think that we left this apart because of the lack of applicability of the PINNs to the project, but they work for other problems like solving some PDEs. 

If that is all right, i can add those small modifications and we could merge this PR.

**Comment by ArnauMiro** [comment]
That would be awesome. When you're done please add me as a reviewer and we start iterating.

**Comment by DavidRamosArchilla** [comment]
I've added a test for the PINN too. However, the test does not check if the pde residual is under a specific threshold or something like that, it just makes sure that training, saving and loading work.

File: requirements_NN.txt [hunk 1]
```diff
@@ -2,4 +2,5 @@ torch>=2.3.0
 torchvision
 torchsummary
 tensorflow
-optuna
\ No newline at end of file
+optuna
+tqdm
```
Comment (ArnauMiro) [inline requirements_NN.txt:6]: extra dependency

**Review by ArnauMiro — CHANGES_REQUESTED** [review]
Why the PINN test doesn't check for the residual of the PDE? How can then we be sure that the PINN is actually working properly?

**Comment by DavidRamosArchilla** [comment]
> Why the PINN test doesn't check for the residual of the PDE? How can then we be sure that the PINN is actually working properly?

I don't know how to implement that on this testsuite, but we could check if the residual is lower that 10⁻³ for example

File: pyLOM/NN/architectures/pinn.py [hunk 2]
```diff
@@ -0,0 +1,541 @@
...
+                desc = f"Epoch {epoch+1}/{epochs} Iteration {closure.iteration}. Pde loss: {loss_from_pde:.4e}" + extended_desc
+                pbar.set_description(desc)
+
+            closure.iteration += 1
+            return loss
+        
+        try:
+            self.model.train()
+            desc = f"Epoch 1/{epochs} Iteration 0."
+            pbar = tqdm(range(epochs), desc=desc)
```
[the remark this answers is not in the document]
Comment (DavidRamosArchilla) [reply pyLOM/NN/architectures/pinn.py:223]: Okay, i'll change that

File: pyLOM/NN/architectures/pinn.py [hunk 3]
```diff
@@ -212,16 +212,15 @@ def closure():
                 if 'test_loss' in logs and len(logs['test_loss']) > 0:
                     extended_desc += f", test loss: {logs['test_loss'][-1]:.4e}"
                 desc = f"Epoch {epoch+1}/{epochs} Iteration {closure.iteration}. Pde loss: {loss_from_pde:.4e}" + extended_desc
-                pbar.set_description(desc)
+                pprint(0, desc)
 
             closure.iteration += 1
             return loss
         
         try:
```
Comment (ArnauMiro) [inline pyLOM/NN/architectures/pinn.py:220]: I'm not sure about this try, except clause. They're very slow and can mask other issues. Is this absolutely necessary?

**Review by ArnauMiro — CHANGES_REQUESTED** [review]
I'll fix the part of the testsuite relating to the comparison with a reference.

File: pyLOM/NN/architectures/pinn.py (hunk 3 above)
Comment (DavidRamosArchilla) [reply pyLOM/NN/architectures/pinn.py:220]: that try except was meant to be able to stop the training manually in both, notebook and command line executions. I don't think that this particular try except will slow down the training, but if you want, i can remove it.

Comment (ArnauMiro) [reply pyLOM/NN/architectures/pinn.py:220]: The idea of pyLOM is to run on headless servers so a manual stop won't be possible usually

Comment (DavidRamosArchilla) [reply pyLOM/NN/architectures/pinn.py:220]: okey, i'll remove it then

**Review by ArnauMiro — CHANGES_REQUESTED** [review]
Pending testsuite fix by @ArnauMiro

**Review by ArnauMiro — APPROVED** [review]
